### PR TITLE
Fix: String#tr String#tr! String#tr_s String#tr_s!

### LINKED ISSUE #765: String#tr is broken implementation for character ranges and other regexp patterns
## ruby/spec

```console
$ bundle exec rake spec
cargo run -q -- enforced-specs.toml
artichoke 0.1.0-pre.0 (2021-01-02 revision 3929) [x86_64-unknown-linux-gnu]

String#tr_s: NNNNXNSNNNNN
String#tr_s!: NNNX
String#tr: .XXXXX.XXS...
String#tr!: XXXX
String.try_convert: ........

Passed 11, skipped 2, not implemented 14, failed 14 specs.

An exception occurred during: Mock.verify_count
String#tr_s tries to convert from_str and to_str to strings using to_str
SpecExpectationNotMetError: Mock 'ab' expected to receive to_str(:any_args) exactly 1 times
but received it 0 times

(eval):653:in each
(eval):140:in each
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_s_spec.rb:104
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr_s! raises a FrozenError if self is frozen
RuntimeError: frozen string

(eval):713:in tr_s!
/src/lib/core/string/tr_s_spec.rb:130:in protect
/src/lib/core/string/tr_s_spec.rb:130:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_s_spec.rb:134
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr accepts c1-c2 notation to denote ranges of characters
SpecExpectationNotMetError: Expected "b-zzz" == "ifmmp"
to be truthy but was false

/src/lib/core/string/tr_spec.rb:13:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:99
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr pads to_str with its last char if it is shorter than from_string
SpecExpectationNotMetError: Expected "A-HH." == "HE..."
to be truthy but was false

/src/lib/core/string/tr_spec.rb:21:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:99
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr raises an ArgumentError a descending range in the replacement as containing just the start character
SpecExpectationNotMetError: Expected ArgumentError but no exception was raised ("z-bbb" was returned)

/src/lib/core/string/tr_spec.rb:25:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:99
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr raises an ArgumentError a descending range in the source as empty
RegexpError: empty range in char class

(eval):694:in tr
/src/lib/core/string/tr_spec.rb:29:in protect
/src/lib/core/string/tr_spec.rb:29:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:99
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr translates chars not in from_string when it starts with a ^
SpecExpectationNotMetError: Expected "ca345cccc" == "cc345cccc"
to be truthy but was false

/src/lib/core/string/tr_spec.rb:34:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:99
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr tries to convert from_str and to_str to strings using to_str
NoMethodError: undefined method 'length'

(eval):692:in tr
/src/lib/core/string/tr_spec.rb:57:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:99
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

An exception occurred during: Mock.verify_count
String#tr tries to convert from_str and to_str to strings using to_str
SpecExpectationNotMetError: Mock 'ab' expected to receive to_str(:any_args) exactly 1 times
but received it 0 times

(eval):653:in each
(eval):140:in each
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:99
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr returns subclass instances when called on a subclass
SpecExpectationNotMetError: Expected "hallo" (String) to be an instance of StringSpecs::MyString

/src/lib/core/string/tr_spec.rb:61:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:99
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr! modifies self in place
SpecExpectationNotMetError: Expected "ab22212hijklmnopqR" == "ab12222hijklmnopqR"
to be truthy but was false

/src/lib/core/string/tr_spec.rb:104:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:129
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr! returns nil if no modification was made
RegexpError: empty char-class

(eval):694:in tr
(eval):702:in tr!
/src/lib/core/string/tr_spec.rb:111:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:129
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr! does not modify self if from_str is empty
RegexpError: empty char-class

(eval):694:in tr
(eval):702:in tr!
/src/lib/core/string/tr_spec.rb:117:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:129
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

String#tr! raises a FrozenError if self is frozen
RuntimeError: frozen string

(eval):700:in tr!
/src/lib/core/string/tr_spec.rb:125:in protect
/src/lib/core/string/tr_spec.rb:125:in protect
(eval):193:in all?
(eval):653:in each
/src/lib/core/string/tr_spec.rb:129
(eval):653:in each
/src/lib/test/spec_runner:186:in run_specs

Passed 11, skipped 2, not implemented 14, failed 14 specs.
rake aborted!
Command failed with status (1): [cargo run -q -- enforced-specs.toml...]
/home/lopopolo/dev/artichoke/artichoke/Rakefile:126:in `block (2 levels) in <top (required)>'
/home/lopopolo/dev/artichoke/artichoke/Rakefile:125:in `chdir'
/home/lopopolo/dev/artichoke/artichoke/Rakefile:125:in `block in <top (required)>'
Tasks: TOP => spec
(See full trace by running task with --trace)

```

### `String#tr`:

https://github.com/artichoke/artichoke/blob/939fffc015255736e4f80fee422efdc88af2a620/artichoke-backend/src/extn/core/string/string.rb#L689-L704

### `String#gsub`:

https://github.com/artichoke/artichoke/blob/939fffc015255736e4f80fee422efdc88af2a620/artichoke-backend/src/extn/core/string/string.rb#L406-L441

### PR DESCRIPTION
**Pull request by ekroon**
Closes #765

### CODE REVIEW

**Comment by lopopolo** [comment]
Hi @ekroon thanks for the PR. I haven't gotten a chance to review this yet, but I noticed the _Audit_ workflow failing for a silly reason so, I put up https://github.com/artichoke/artichoke/pull/1013.

If you rebase to pick up https://github.com/artichoke/artichoke/commit/ac3086fc77aa6c773791c506e774fc629facefc9, the build should turn all green.

**Comment by ekroon** [comment]
Hi @lopopolo this is ready for review now. I refactored some code duplication from first version and a superfluous call to `to_a` after `chars`. I would like to replace the call to `chars.map` to `each_char`, but `each_char` is not implemented yet, so I left it like this.

File: artichoke-backend/src/extn/core/string/string.rb [hunk 1]
```diff
@@ -785,4 +776,67 @@ def valid_encoding?
...
+      else
+        arr.push(str_a[i])
+        i += 1
+      end
+    end
+    arr
+  end
+
+  def tr_compose_str(container, from, to, skip_double, index_check, retrieve_value)
+    last_c = ''
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/string/string.rb]:
```suggestion
    previous_char = ''
```

File: artichoke-backend/src/extn/core/string/string.rb [hunk 2]
```diff
@@ -785,4 +776,67 @@ def valid_encoding?
...
+                             ''
+                           else
+                             new_c
+                           end
+                     last_c = new_c
+                     res
+                   else
+                     last_c = ''
+                     c
+                   end
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/string/string.rb]: The formatting RuboCop puts here makes this critical section hard to read. I'd also like to expand the variable names to make this easier to follow. Can we rewrite this to something like:

```suggestion
      if index_check.call(index)
        proposed_replacement_char = retrieve_value.call(to, index)
        replacement_char = proposed_replacement_char
        if skip_double && proposed_replacement_char == previous_char
          replacement_char = ''
        end
        previous_char = proposed_replacement_char
        container << replacement_char
      else
        previous_char = ''
        container << c
      end
```

File: artichoke-backend/src/extn/core/string/string.rb [hunk 3]
```diff
@@ -785,4 +776,67 @@ def valid_encoding?
     # NOOP and is here for compatibility.
     true
   end
+
+  private
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/string/string.rb]: For helper methods like this written in Ruby, I've tended to move them out of the core classes into an `Artichoke` helper module, like this one for `Array#inspect`:

https://github.com/artichoke/artichoke/blob/6b458183f932ef8bf96d5683fe4c5366b948e421/artichoke-backend/src/extn/core/array/array.rb#L3-L11

Can we move these `tr` helper methods into an `Artichoke::String` module located at the top of this source file?

We'd then delegate to these methods in the `String#tr` and friends impls:

https://github.com/artichoke/artichoke/blob/6b458183f932ef8bf96d5683fe4c5366b948e421/artichoke-backend/src/extn/core/array/array.rb#L916-L933

Comment (lopopolo) [reply artichoke-backend/src/extn/core/string/string.rb]: Also, as a current implementation limitation of Artichoke, `private` without a method name has no effect. To make a method private, one must call `private :tr_expand_str` after the method definition.

File: artichoke-backend/src/extn/core/string/string.rb [hunk 4]
```diff
@@ -785,4 +776,67 @@ def valid_encoding?
...
+      from = tr_expand_str(from_str)
+      to = tr_expand_str(to_str)
+      tr_compose_str(result, from, to, skip_double, ->(index) { index.nil? }, ->(lookup, _) { lookup.last })
+    else
+      from = tr_expand_str(from_str)
+      to = tr_expand_str(to_str)
+      tr_compose_str(result, from, to, skip_double, ->(index) { !index.nil? }, ->(lookup, index) { lookup[index] || lookup.last || '' })
+    end
+    result
+  end
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/string/string.rb]: This is super easy to follow. I like how you've factored the implementation into separate _config_ and _execute_ steps.

**Review by lopopolo — CHANGES_REQUESTED** [review]
Hi @ekroon this looks great. I've asked for a couple of changes:

- Can we move the private methods out of the `String` scope to an `Artichoke::String` class? Despite being private methods, these change the prototype of `String` in an incompatible way if we wish to support Ruby users monkey patching `String` arbitrarily without breaking `String#tr` and friends.
- I've taken a stab at refactoring `tr_compose_str` to make it a bit easier to read.

Thanks for taking the time to work through this. The implementation is super clean. I'm excited to increase the ruby/spec coverage with this PR! 💎 

Let me know what you think.

**Comment by ekroon** [comment]
@lopopolo Thanks for the review. I implemented your proposed changes. I added a separated commit but you might want them squashed before merging?

**Review by lopopolo — APPROVED** [review]
Thanks for making these changes 🚀 💎

**Pull request merged**